### PR TITLE
feat: added husky for pre-commit

### LINKED ISSUE #1746: support pre-commit
# Feature request

## Please describe your feature

Hi, I noticed that there have something that could be automatically done before committing:

- [ ] Auto prettier;
- [ ] Auto check commit message;

## Describe the solution you'd like

I would prefer using [husky](https://www.npmjs.com/package/husky) to do this.

## Additional context

Due to we put files under the `web` folder (non-root), husky doesn't support this way months ago, please check if it supports it now.

### PR DESCRIPTION
**Pull request by iamayushdas**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

- [x] auto prettier
- [ ] auto-check commit message

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [x] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Added auto prettier via husky 

**Related issues**

fix/resolve #1746 #1900 

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: web/src/global.tsx [hunk 1]
```diff
@@ -25,8 +25,8 @@ import defaultSettings from '../config/defaultSettings';
 
 // see https://github.com/suren-atoyan/monaco-react/issues/168
 loader.config({ paths: { vs: '/monaco-editor/min/vs' } });
-if (getLocale() === 'zh-CN'){
-  loader.config({ "vs/nls": { availableLanguages: { "*": 'zh-cn' } } });
+if (getLocale() === 'zh-CN') {
```
Comment (iamayushdas) [inline web/src/global.tsx:28]: staged this file to check the husky

**Comment by iamayushdas** [comment]
cc @juzhiyuan @guoqqqi @liuxiran

**Comment by juzhiyuan** [comment]
Hi, I just checked this PR, and your solution is creating a Husky under the root folder, actually, we only want to format the `web/` folder 🤔 

Do we have some reference or examples using this way?

**Comment by iamayushdas** [comment]
> Hi, I just checked this PR, and your solution is creating a Husky under the root folder, actually, we only want to format the `web/` folder 🤔 
> 
> 
> 
> Do we have some reference or examples using this way?

https://typicode.github.io/husky/#/?id=custom-directory

**Comment by juzhiyuan** [comment]
that's fine, but this PR put package.json under the root folder, right?

I think it should be at the `web` folder just like this example:

https://typicode.github.io/husky/#/?id=custom-directory

```json
// package.json
{
  "scripts": {
    "prepare": "cd .. && husky install front/.husky"
  }
}
```

**Comment by juzhiyuan** [comment]
image

**Comment by iamayushdas** [comment]
If we install something in some package (eg: husky) it will create the package.json, and also IMO that package.json with script consisting in it is the same package.json that will produce while installing husky

**Comment by iamayushdas** [comment]
Okay okay, i guess i understood you now, i am working on it as you want

**Comment by iamayushdas** [comment]
cc @juzhiyuan check it again please

**Review by juzhiyuan — APPROVED** [review]

**Comment by juzhiyuan** [comment]
cc @bzp2010 @liuxiran @LiteSun

**Review by guoqqqi — APPROVED** [review]

**Comment by liuxiran** [comment]
I tried in my local env, and found that when I tried to commit a `js` file e.g. cypress test files or `.eslintrc.js`, an error will occur:

image

That is because we did not put these js files into `tsconfig.json` include list, or we can also choose to ignore them

I suggest to put `cypress` into include list, as for other config files like `.eslintrc.js`, we can put them into `.eslintignore` file

what do you think?

cc @juzhiyuan @guoqqqi @iamayushdas @LiteSun

**Comment by iamayushdas** [comment]
I agree @liuxiran , it is important for tests, thanks for review

**Comment by LiteSun** [comment]
Agree to put cypress into include list.

**Comment by iamayushdas** [comment]
@liuxiran @LiteSun check once again, is it fine now?

**Comment by iamayushdas** [comment]
search_ route.spec.js test is failing

**Comment by guoqqqi** [comment]
I think we should add some assertions to these open drawer tests to make them as stable as possible.

File: web/.eslintignore [hunk 2]
```diff
@@ -1,4 +1,5 @@
 /lambda/
 /scripts
 /config
-.history
\ No newline at end of file
+.history
+.eslintrc.js
```
Comment (liuxiran) [inline web/.eslintignore:5]: empty line

**Comment by liuxiran** [comment]
image

we can try to import `uuid` which is already one of our dependencies to  generate the `uid` for the file

**Comment by iamayushdas** [comment]
cc 

> image
> 
> we can try to import `uuid` which is already one of our dependencies to generate the `uid` for the file

we can add this in later PR, is that cool for you

**Comment by iamayushdas** [comment]
cc @juzhiyuan @liuxiran @LiteSun @guoqqqi need you review again and lets get this feature merged,

**Comment by liuxiran** [comment]
> we can add this in later PR, is that cool for you

👏  so cool ~, that's a deal

**Review by liuxiran — APPROVED** [review]

**Review by juzhiyuan — APPROVED** [review]

**Comment by iamayushdas** [comment]
@nic-chen can i have your review here

File: web/src/constants.ts [hunk 3]
```diff
@@ -28,6 +28,7 @@ export const codeMessage = {
   422: '当创建一个对象时，发生一个验证错误。',
   500: '服务器发生错误，请检查服务器。',
   502: '网关错误。',
+
```
Comment (nic-chen) [inline web/src/constants.ts:31]: why change this?

Comment (iamayushdas) [reply web/src/constants.ts:31]: From that i was just testing husky, it will not harm any code

Comment (nic-chen) [reply web/src/constants.ts:31]: Yes, but please don't make unnecessary changes.

**Comment by iamayushdas** [comment]
cc @LiteSun

**Review by LiteSun — APPROVED** [review]

**Pull request merged**